Use GitHub API for basic checks/info during publishing

mosteo opened the pull request:
Due to the problem of regular users not being able to create branches in our community index, this is somewhat mitigated in this PR by:

- Requesting the user GitHub username
- Checking that the user exists
- Checking the user has forked the index repo
- Checking the fork contains the base branch for PRs (otherwise the upload URL fails).

These incremental checks allow us to provide some guidance at each failed check on what to do to continue.

To implement this quickly I've used the GitHub REST API with calls that don't require authorization or parsing the resulting JSON. Mid-term we can explore whether it's easier to use this API with an authorization token, or the `gh` command-line official tool, to automatically perform these steps, create releases, upload files, and create the PR.

mosteo committed:
Use GitHub API for basic checks/info during publish

Fabien-Chouteau commented on src/alire/alire-publish.adb:
```diff
@@ -522,6 +539,80 @@ package body Alire.Publish is
+
+      --  User has an account
+
+      if not Config.Defined (Config.Keys.User_Github_Login) then
```
You can use `Query_Config.User_GitHub_Login` here it will ask the user for it's login or take it from the config.

mosteo replied:
Yes, I use it later, I make this check to print a bit of extra context before asking for the first time.

Fabien-Chouteau replied:
I see for this one. The other one above I don't understand.

Fabien-Chouteau commented on src/alire/alire-utils-user_input-query_config.adb:
```diff
@@ -36,7 +36,7 @@ package body Alire.Utils.User_Input.Query_Config is
    function User_Name return String
    is (Config_Or_Query_String (Config_Key => "user.name",
-                               Question   => "User name",
+                               Question   => "Please enter your user name:",
```
It should probably `Please enter your name:`, there no "user name" for Alire.

mosteo replied:
Right, oof.

mosteo committed:
Code review fix

Fabien-Chouteau commented on src/alire/alire-publish.adb:
```diff
@@ -255,12 +256,28 @@ package body Alire.Publish is
+         --  Show the upload URL in normal circumstances, or a more generic
+         --  message otherwise (when lacking a github login).
+
+         if Config.Defined (Config.Keys.User_Github_Login) then
```
Why is there an alternative here? The user need a GitHub account to contribute to the index anyway.

mosteo replied:
This one is because when `--force` is in effect we skip the checks about github, so it's possible to reach this stage without a configured github login, if you never run `alr init` before.

I did it like this to allow our tests to succeed locally, and to allow people that wants to publish to other indexes besides the community one, to be able to get the manifest without requiring a real github account. But then it's true the message for that case might not be appropriate.

mosteo commented:
@reznikmm, this PR addresses the problem you found with creating PRs against the community index. It would be great if you had the time to give it a try and see whether you detect something amiss.

reznikmm commented:
It looks fine:

```
Publishing assistant: step 2 of 6: Verify GitHub infrastructure
ⓘ Publishing to the community index requires a GitHub account.
Please enter your GitHub login: (default: 'github-username')
> reznikmm
✓ User has forked the community repository
✓ User's fork contains base branch: devel-0.4
...
✓ Your index manifest file has been generated at /tmp/axmpp/alire/releases/axmpp-1.0.0.toml
ⓘ Please create a pull request against the community index at https://github.com/alire-project/alire-index including this file at index/ax/axmpp
```

mosteo commented:
> It looks fine:

Mmm actually the link you are provided with at the end is wrong, as it should point to your fork. I can reproduce, so thanks!

mosteo committed:
Fix: new config values unavailable during same run

New config entries were stored on disk but not added to the memory map. As the
Config.Set subprogram isn't aware of the level of configuration being set,
there is no direct way to both store to disk and to memory; instead the
solution is that the configuration is reloaded every time. This is a bit of
overkill and a candidate for later optimization.

mosteo commented:
@Fabien-Chouteau , can you please check the last commit that fixes the config storage issue? As I'm not 100% familiar with the config internals.

zertovitch commented:
Some builds' fails can be explained: it is due to the lack of support for 128-bit integers.

zertovitch commented:
The reliance on 128-bit integers makes HAC too restrictive in respect to the compiler choice. I'll use something else for that computation...
